working/1426-extension-types/feature-specification.md

```diff
@@ -0,0 +1,752 @@
+
+
+## Enhancements
+
+
+### Non-object entities
+
+If we introduce any non-object entities in Dart (that is, entities that
+cannot be assigned to a variable of type `Object?`, e.g., external C /
+JavaScript / ... entities, or non-boxed tuples, etc), then we may wish to
+allow for extension types whose on-type is a non-object type.
+
+This should not cause any particular problems: If the on-type is a
+non-object type then the extension type will not be a subtype of `Object`.
+
+
+## Discussion
+
+It would be possible to reify extension types when they occur as type
+arguments of a generic type.
+
+This might help ensuring that the associated discipline of the extension
+type is applied to the elements in, say, a list, even in the case where
+that list is obtained under the type `dynamic`, and a type test or type
+cast is used to confirm that it is a `List<U>` where `U` is an extension
+type.
+
+However, this presumably implies that the cast to a plain `List<T>` where
+`T` is the on-type corresponding to `U` should fail; otherwise the
+protection against accessing the elements using the underlying on-type will
+easily be violated. Moreover, even if we do make this cast fail then we
+could cast each element in the list to `T`, thus still accessing the
+elements using the on-type rather than the more disciplined extension type
+`U`.
+
+We cannot avoid the latter if there is no run-time representation of the
+extension type in the elements in the list, and that is assumed here: For
+example, if we have an instance of `int`, and it is accessed as `extension
+MyInt on int`, the dynamic representation will be a plain `int`, and not
+some wrapped entity that contains information that this particular `int` is
+viewed as a `MyInt`. It seems somewhat inconsistent if we maintain that a
+`List<MyInt>` cannot be viewed as a `List<int>`, but a `MyInt` can be
+viewed as an `int`.
+
+As for promotion, we could consider "promoting to a supertype" when that
+type is an extension type: Assume that `U` is an extension type with
+on-type `T`, and the type of a promotable local variable `x` is `T` or a
+subtype thereof; `x is U` could then _demote_ `x` to have type `U`, even
+though `is` tests normally do not demote. The rationale would be that the
+treatment of `x` as a `U` is conceptually more "informative" and "strict"
+than the treatment of `x` as a `T`, which makes it somewhat similar to a
+downcast.
+
+Note that we can use extension types to handle `void`:
+
+```dart
+extension void on Object? {}
+```
+
+This means that `void` is an "epsilon-supertype" of all top types (it's a
+proper supertype, but just a little bit). It is also a subtype of
+`Object?`, of course, so that creates an equivalence class of "equal" types
+spelled differently. That's a well-known concept today, so we can handle
+that (and it corresponds to the dynamic semantics).
+
+This approach does not admit any member accesses for a receiver of type
```

Comment:
We allow you to can call `Object` members on extension types, so not entirely correct.

Reply:
I added a `hide` clause to fix that.